### PR TITLE
make it work with ruby 3.1 by upgrading thin to the latest version …

### PR DESCRIPTION
**Pull request by Wanderfalke**
…and replacing skinny with faye-websocket.

also allow mailcatcher process to be stopped with SIGINT / CTRL+C

### CODE REVIEW

**Comment by Vincent14** [comment]
Hi @sj26 or @tagliala @westonganger  could you have a look please? <3

**Comment by westonganger** [comment]
Unfortunately I am not a maintainer of this project. I think I reached out before. But if more maintainership help is needed I am willing.

**Comment by sj26** [comment]
Thanks for the contribution! But there is already a conflicting stategy for upgrading thin and websockets etc over here:
https://github.com/sj26/mailcatcher/pull/467

The blocker is testing, including adding more feature spec coverage. If you're keen to contribute, looking at the surface area of mailcatcher and adding feature specs would be a great contribution.

**Comment by sj26** [comment]
> allow mailcatcher process to be stopped with SIGINT / CTRL+C

This should already be working:

image

**Comment by sj26** [comment]
Reading this again, it's a pretty small intervention so might be easier to land. I'll do a review and maybe we can land it.

File: lib/mail_catcher.rb [hunk 1]
```diff
@@ -178,6 +178,12 @@ def run! options=nil
 
     # One EventMachine loop...
     EventMachine.run do
+      trap("INT") do
+        EventMachine.add_timer(0) do
```
Comment (sj26) [inline lib/mail_catcher.rb:182]: I think EventMachine should already be quitting cleanly. But, if not, I think for this line you want:

```suggestion
        EventMachine.next_tick do
```

File: lib/mail_catcher/web/application.rb [hunk 2]
```diff
@@ -5,14 +5,12 @@
 require "uri"
 
 require "sinatra"
-require "skinny"
+require 'faye/websocket'
```
Comment (sj26) [inline lib/mail_catcher/web/application.rb:8]: Please follow the conventions in the surrounds:

```suggestion
require "faye/websocket"
```

File: lib/mail_catcher/web/application.rb [hunk 3]
```diff
@@ -5,14 +5,12 @@
...
 
 require "mail_catcher/bus"
 require "mail_catcher/mail"
 
-class Sinatra::Request
-  include Skinny::Helpers
-end
+Faye::WebSocket.load_adapter('thin')
```
Comment (sj26) [inline lib/mail_catcher/web/application.rb:13]:
```suggestion
Faye::WebSocket.load_adapter("thin")
```

File: mailcatcher.gemspec [hunk 4]
```diff
@@ -37,8 +37,9 @@ Gem::Specification.new do |s|
   s.add_dependency "rack", "~> 1.5"
   s.add_dependency "sinatra", "~> 1.2"
   s.add_dependency "sqlite3", "~> 1.3"
-  s.add_dependency "thin", "~> 1.5.0"
-  s.add_dependency "skinny", "~> 0.2.3"
+  s.add_dependency "thin", "~> 1.8.0"
+  s.add_dependency "net-smtp", "~> 0.3.0"
```
Comment (sj26) [inline mailcatcher.gemspec:41]: Why was net-smtp added here?

Comment (tagliala) [reply mailcatcher.gemspec:41]: > net-smtp

`net-smtp` has been extracted from the default gems in Ruby 3.1

Ref: 
- rails/rails#44083
- https://www.ruby-lang.org/en/news/2021/12/25/ruby-3-1-0-released/

File: lib/mail_catcher.rb (hunk 1 above)
Comment (Wanderfalke) [reply lib/mail_catcher.rb]: You're right. The magic was to pass _signals: false_ to thin.

File: lib/mail_catcher/web/application.rb (hunk 2 above)
Comment (Wanderfalke) [reply lib/mail_catcher/web/application.rb]: fixed

File: lib/mail_catcher/web/application.rb (hunk 3 above)
Comment (Wanderfalke) [reply lib/mail_catcher/web/application.rb]: fixed

File: mailcatcher.gemspec (hunk 4 above)
Comment (Wanderfalke) [reply mailcatcher.gemspec:41]: As already mentioned by @tagliala, as of Ruby 3.1, net-smtp was moved to bundled gems.

**Comment by Wanderfalke** [comment]
@sj26 Thank you for taking time to review my PR. I addressed your comments. Please have another look when you can.

File: lib/mail_catcher.rb (hunk 1 above)
Comment (sj26) [reply lib/mail_catcher.rb]: But now thin doesn't exit cleanly?

Sorry, by "EventMachine should already be quitting cleanly" I meant that thin's signal handler is enough to shut down the reactor including draining thin's connections cleanly. Did that break? Is that why you made this change initially?

File: mailcatcher.gemspec (hunk 4 above)
Comment (sj26) [reply mailcatcher.gemspec:41]: Ah, I see. Thanks.

File: lib/mail_catcher.rb (hunk 1 above)
Comment (sj26) [reply lib/mail_catcher.rb]: Ctrl+C now does this for me:

```
 ▲ ~/Developer/mailcatcher bundle exec mailcatcher --foreground                                                                               Wanderfalke-main 9m ● ⬡
Starting MailCatcher v0.8.1
==> smtp://127.0.0.1:1025
/opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/net/protocol.rb:66: warning: already initialized constant Net::ProtocRetryError
/opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/net-protocol-0.1.2/lib/net/protocol.rb:68: warning: previous definition of ProtocRetryError was here
/opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/net/protocol.rb:206: warning: already initialized constant Net::BufferedIO::BUFSIZE
/opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/net-protocol-0.1.2/lib/net/protocol.rb:208: warning: previous definition of BUFSIZE was here
/opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/net/protocol.rb:503: warning: already initialized constant Net::NetPrivate::Socket
/opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/net-protocol-0.1.2/lib/net/protocol.rb:504: warning: previous definition of Socket was here
==> http://127.0.0.1:1080
^CTraceback (most recent call last):
	20: from /opt/rbenv/versions/2.7.5/bin/bundle:23:in `<main>'
	19: from /opt/rbenv/versions/2.7.5/bin/bundle:23:in `load'
	18: from /opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/libexec/bundle:34:in `<top (required)>'
	17: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/friendly_errors.rb:123:in `with_friendly_errors'
	16: from /opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/libexec/bundle:46:in `block in <top (required)>'
	15: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/cli.rb:24:in `start'
	14: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
	13: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/cli.rb:30:in `dispatch'
	12: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
	11: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	10: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	 9: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/cli.rb:476:in `exec'
	 8: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/cli/exec.rb:28:in `run'
	 7: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/cli/exec.rb:63:in `kernel_load'
	 6: from /opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/bundler/cli/exec.rb:63:in `load'
	 5: from /opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/bin/mailcatcher:23:in `<top (required)>'
	 4: from /opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/bin/mailcatcher:23:in `load'
	 3: from /Users/sj26/Developer/mailcatcher/bin/mailcatcher:6:in `<top (required)>'
	 2: from /Users/sj26/Developer/mailcatcher/lib/mail_catcher.rb:180:in `run!'
	 1: from /opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:193:in `run'
/opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:193:in `run_machine': Interrupt
```

File: mailcatcher.gemspec (hunk 4 above)
Comment (sj26) [reply mailcatcher.gemspec:41]: This spits a lot of warnings during startup for me on Ruby 2.7.5:

```
/opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/net/protocol.rb:66: warning: already initialized constant Net::ProtocRetryError
/opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/net-protocol-0.1.2/lib/net/protocol.rb:68: warning: previous definition of ProtocRetryError was here
/opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/net/protocol.rb:206: warning: already initialized constant Net::BufferedIO::BUFSIZE
/opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/net-protocol-0.1.2/lib/net/protocol.rb:208: warning: previous definition of BUFSIZE was here
/opt/rbenv/versions/2.7.5/lib/ruby/2.7.0/net/protocol.rb:503: warning: already initialized constant Net::NetPrivate::Socket
/opt/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/net-protocol-0.1.2/lib/net/protocol.rb:504: warning: previous definition of Socket was here
```

**Comment by sj26** [comment]
I've added a spec on main which tests that mailcatcher exits cleanly on interrupt, and rebasing this branch to include that spec causes a failure:

```
  1) Quit quits cleanly on Ctrl+C
     Failure/Error: expect(status).to be_exited
       expected `#<Process::Status: pid 56232 SIGINT (signal 2)>.exited?` to be truthy, got false
     # ./spec/quit_spec.rb:41:in `block (2 levels) in <top (required)>'
```

File: lib/mail_catcher.rb (hunk 1 above)
Comment (Wanderfalke) [reply lib/mail_catcher.rb:182]: My initial motivation for this was the fact, that for me on MacOS 12.1 (Intel) CTRL+C / SIGINT did nothing and mailcatcher process can only be stopped with SIGKILL. 
Passing "signals: false" to thin makes SIGINT work, but as you state above also prints stacktrace. Explicitely trapping INT signal in the eventmachine loop makes it exit cleanly.
BTW, your suggestion to use "EventMachine.next_tick" leads to following error: `eventmachine.rb:1117:in "synchronize": can't be called from trap context (ThreadError)`

File: mailcatcher.gemspec (hunk 4 above)
Comment (Wanderfalke) [reply mailcatcher.gemspec:41]: Tried ruby 2.7.5 and I'm getting same warnings. Not sure how to avoid that yet, will keep looking for a solution.

Comment (Wanderfalke) [reply mailcatcher.gemspec:41]: Following this issue: https://github.com/ruby/net-imap/issues/16 , adding net-http as dependency solves "already initialized constant" issues with net-protocol.

**Comment by shaneshort** [comment]
any particular reason this wasn't merged? I'm currently unable to install this on ruby 3.x and I suspect this PR would help.

**Comment by sj26** [comment]
There is a failing spec rebased on main, and not enough testing generally to be sure it hasn't broken anything.

**Comment by sj26** [comment]
Hi folks! Thanks for this work here, especially @Wanderfalke.

Ruby 3.1 support and a switch to faye-websocket are available in [0.9.0.beta1](https://github.com/sj26/mailcatcher/releases/tag/v0.9.0.beta1). I've love some folks to test. You can install the latest pre-release with:

```ruby
gem install --pre mailcatcher
```

I think there are still some quirks. I don't think the websockets send the quit message before shut down. But that's not critical.

I'm keen to upgrade Thin, too, but I'd like to bed in the switch to faye, first. Beyond that I'll see if we can complete the switch to the async gems (#467).

**Comment by larryh** [comment]
I know this is marked as **Closed** (so hopefully I'm not wasting my time typing this), but I wanted to give the maintainers and anyone else who has this problem some feedback...

I'm using Rails 3.1.0, and when I installed and tried to run mailcatcher 0.9.0 this is what I got:
<pre>
Unable to load the EventMachine C extension; To use the pure-ruby reactor, require 'em/pure_ruby'
<internal:/home/larry/.rbenv/versions/3.1.0/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require': libssl.so.1.1: cannot open shared object file: No such file or directory - /home/larry/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/eventmachine-1.0.9.1/lib/rubyeventmachine.so (LoadError)
        from <internal:/home/larry/.rbenv/versions/3.1.0/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from /home/larry/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:8:in `<top (required)>'
        from <internal:/home/larry/.rbenv/versions/3.1.0/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from <internal:/home/larry/.rbenv/versions/3.1.0/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from /home/larry/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mailcatcher-0.8.1/lib/mail_catcher.rb:7:in `<top (required)>'
        from <internal:/home/larry/.rbenv/versions/3.1.0/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
 </pre>

So I ran the following two commands:
<pre>
gem uninstall mailcatcher
gem install --pre mailcatcher
</pre>

And it worked like a charm, specifically:

<pre>
larry@system76-pc:~/RubymineProjects/BackEnd$ mailcatcher
Starting MailCatcher v0.9.1.beta1
==> smtp://127.0.0.1:1025
==> http://127.0.0.1:1080
*** MailCatcher runs as a daemon by default. Go to the web interface to quit.
</pre>

Thanks for the great gem!

**Comment by sj26** [comment]
Sorry @larryh, that's very weird!

That sounds like a problem within eventmachine. Perhaps it linked to an older version of openssl? Are you using homebrew? It might have upgraded libraries beneath your feet.

Reinstalling the gems usually fixes it up for me:

```
gem pristine eventmachine
# or
gem pristine --all
```

**Comment by larryh** [comment]
Hi,
Thanks for the constant feedback - I know you must be busy and it is much appreciated.

A "gem list" command tells me I have Eventmachine 1.2.7, which I believe is the latest version.(From 5 years ago!)  It's not in my (Rails project) Gemfile or even Gemfile.lock. Also, I don't think I ever intentionally installed it, so maybe it just comes along for the ride when I install Mailcatcher. 

Also, I'm not on a Mac, so no homebrew. (I'm on Ubuntu)

I'll just continue to refresh the page to see new emails.

Thanks again,
Larry

**Comment by sj26** [comment]
Heh, Homebrew runs on Linux these days.

But even apt may have bumped your libssl. And apt isn't aware of rubygems, so wouldn't cause dependent gems to be recompiled.

I'd recommend a `gem pristine --all` just to be safe :-)

**Comment by larryh** [comment]
Thanks for the tip. I did a "gem pristine eventmachine" but that didn't do anything; I still have to refresh the page to see the email.

I don't want to try the "all" command because I'm gun-shy of potential unintended consequences. Oftentimes (or at least sometimes) issuing commands that sweeping can result in spending hours just trying to get back to where you were, i.e. "when things worked."

I'll just stick with refreshing the page when I need to.

Thanks for all of your help!

**Pull request closed without merging**